### DISCUSSION TITLE
[NEXT-1147] Scroll position is reset when search params are updated

### DISCUSSION THREAD

**Discussion by benjaminwaterlot** [vercel/next.js#60146] [Help]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Next.js `13.3.5-canary.2`
```

### Link to the code that reproduces this issue

https://codesandbox.io/p/sandbox/modest-gould-h4nlvd?file=%2Fapp%2Fpage.tsx&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A7%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A7%7D%5D

### To Reproduce

* Click the counter to update client state. 
* Scroll a little.
* Click the second button to update search params.
* The client state is preserved, but the scroll position is lost.

### Context

In Next <= 13.2.4, updating search params was working as intended : client state was kept, and scroll position was kept too.

In Next 13.2.5, a regression made the client components unmount and remount when search params were updated.

@feedthejim fixed that unmounting in https://github.com/vercel/next.js/pull/49047 (it's testable on `13.3.5-canary.2`), but **there is still an issue now: scroll position is lost on search params updates**.

**Problem**: persisting state in search params is very important with the App router - that's a clean way for client components to request updated data from the RSC.

### Reproduction Codesandbox

Here's a codesandbox reproducing the bug:

https://codesandbox.io/p/sandbox/modest-gould-h4nlvd?file=%2Fapp%2Fpage.tsx&selection=%5B%7B%22endColumn%22%3A1%2C%22endLineNumber%22%3A7%2C%22startColumn%22%3A1%2C%22startLineNumber%22%3A7%7D%5D

### In video

https://user-images.githubusercontent.com/32556322/235664370-77bc7e4a-daaa-4615-9db7-094640873c33.mp4

### A real-world example of the problem

_Previously, the "name for the fund", kept in client state, was reset when updating the sliders. It's now better, but the scroll position reset is still awkward._

https://user-images.githubusercontent.com/32556322/235666855-98472456-121a-4f80-a80d-c4e1eda70aaa.mp4

### Expected Behavior

When updating search params, the scroll position should be preserved.

### Which browser are you using? (if relevant)

Latest stable Google Chrome

<sub>[NEXT-1147](https://linear.app/vercel/issue/NEXT-1147/scroll-position-is-reset-when-search-params-are-updated)</sub>

**Comment by ghost** [comment]
Looking forward to this getting fixed :-)

**Comment by aaron5670** [comment]
I experience the same issue... I thought it was a bug in my code. 😆

**Comment by adrianhajdin** [comment]
I'm experiencing the same bug. The website scrolls to the top every time the URL parameters change. 

I saw @timneutkens's [tweet](https://twitter.com/timneutkens/status/1645138528781709314?s=20) on the topic.

https://github.com/vercel/next.js/assets/24898559/dbbdf535-abaf-481c-82a8-9d6f0149b362

**Comment by Tony0205** [comment]
I'm having the same problem... 
And it seems that the page still remounts when the search param is changed.

My nextJS version is `13.4.5-canary.0`

It seems to be fine when I use English, but when I use Korean, it keeps reloading.

**Comment by ghost** [comment]
@adrianhajdin I'm having the same issue...

**Comment by YoussefKababe** [comment]
Facing the same issue... First I would like to mention that I think this is a correct behavior by default, there nothing wrong with it; however, it would be awesome to have the option to persist scroll position indeed.

I came up with a quick workaround to solve this temporarily until there's an official fix:

```javascript
'use client'
import { useCallback, useEffect } from 'react'
import { useRouter, usePathname, useSearchParams } from 'next/navigation'

export default function ExampleClientComponent(): JSX.Element {
  const router = useRouter()
  const pathname = usePathname()
  const searchParams = useSearchParams()

  useEffect(() => {
    // Retrieve scrollY value from localStorage after routing
    const persistentScroll = localStorage.getItem('persistentScroll')
    if (persistentScroll === null) return

    // Restore the window's scroll position
    window.scrollTo({ top: Number(persistentScroll) })
   
    // Remove scrollY from localStorage after restoring the scroll position
    // This hook will run before and after routing happens so this check is
    // here to make we don't delete scrollY before routing
    if (Number(persistentScroll) === window.scrollY)
      localStorage.removeItem('persistentScroll')
  }, [searchParams])

  const setSearchParam = useCallback(
    (key: string, value: string) => {
      const currentParams = searchParams.toString()
      const params = new URLSearchParams(currentParams)

      params.set(key, value)
      // If search params are still the same there's no need to do anything
      if (currentParams === params.toString()) return

      // Save current scrollY value to localStorage before pushing the new route
      localStorage.setItem('persistentScroll', window.scrollY.toString())
      router.push(`${pathname}?${params.toString()}`)
    },
    [searchParams, pathname, router],
  )

  return (
    <button
      onClick={() => {
        setSearchParam('exampleParam', '9999')
      }}
    >
      Update search params
    </button>
  )
}

```

Depending on your application's logic, you can package this into a React hook and reuse it in all components that should update the search params.

**Comment by GBrachetta** [comment]
@adrianhajdin I got it working by using `"next": "13.2.4"` in `package.json` (and obviously running `npm i` afterwards) and adding 

```js
  experimental: {
    appDir: true,
  },
```

in `next.config.js`.

At least this works fine while keeping server side rendering until they fix the bug.
This might be better than converting the component to client side rendering.

**Comment by AdiAkhileshSingh15** [comment]
@GBrachetta it won't work when deployed i guess.

So, i would suggest people to use that method of restoring the scroll position as shown by @YoussefKababe , that will work in the deployed version as well, though it seems to make a page reload or smthng, but as @GBrachetta said, atleast it's better than converting the component to client side rendering, till the bug's fixed.

**Comment by jeeknowme** [comment]
> I'm experiencing the same bug. The website scrolls to the top every time the URL parameters change.
> 
> I saw @timneutkens's [tweet](https://twitter.com/timneutkens/status/1645138528781709314?s=20) on the topic.
> 
>  2023-05-27.09-58-58.mp4

What i have done in the project I followed from you is that I've just put id="searchbar" in the form then append #searchbar in 
    router.push(`${newPathname}#searchbar`)
    so that every changes in url, it directly goes in the dropdown, just having a slight navigation  after.
    Although this still doesn't resolve the issue, I can still use server side rendering from this.

**Comment by lucasoz** [comment]
Same issue in `"next": "13.4.6"`

**Comment by kevinbateman** [comment]
I'm seeing lots of router fixes in `13.4.6` and `13.4.7-canary*` ref. [https://github.com/vercel/next.js/releases](https://github.com/vercel/next.js/releases), but no notes addressing this issue.

Anyone know if a fix for this is in the works anytime soon?

**Comment by abhishekmardiya** [comment]
I'm encountering a similar issue where the webpage consistently jumps to the top whenever there is a modification to the URL parameters.

**Comment by malburo** [comment]
same issue :(

**Comment by amanchev** [comment]
Same issue.

**Comment by ginabeki** [comment]
Same issue :(

**Comment by weronikadominiak** [comment]
Same issue.

**Comment by dihan48** [comment]
Same issue.

**Comment by e2goon** [comment]
https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#scrolling-to-an-id

![image](https://github.com/vercel/next.js/assets/45934/b24f1a24-ed3f-4d17-9ac3-cbbf96cda216)

```tsx
<Link href="/#hashid" scroll={false}>
  Scroll to specific id.
</Link>
```

This code does not work with `scroll={false}`.
Can the scroll={false} option universally prevent scrolling to the top, not just with hash routing?

**Comment by kusalkalinga00** [comment]
same issue

**Comment by steadyGuy** [comment]
This issue is really frustrating. I don't want to downgrade my next.js to 13.2.4 just to resolve this bug, or else I will get other unresolved problems.

**Comment by aaron5670** [comment]
Same issue, really annoying and a bit blocking.. 🙄

**Comment by GBDev13** [comment]
same issue 🥲

**Comment by mylesthedev** [comment]
Same issue on `13.4.7`. Seems avoidable if the parent route/page doesn't have a wrapping element.
For example if we wrap the route in <></> scroll position is kept weirdly enough.

**Comment by merwinpoulose** [comment]
same issue 🥲

**Comment by shuzipai** [comment]
Similar issues. When using the APP mode, whether using links or router. push or other methods, it will result in a hard refresh of the current page and loss of many operations on the current page. I hope that users can keep the URL and page synchronized when performing some operations, so that the effect obtained when refreshing the page will be consistent with the current operation

**Comment by VladSturm** [comment]
same issue

**Comment by bagedaddy** [comment]
Same issue :(

**Comment by YevhenSoldatov1998** [comment]
Same issue(

**Comment by mkhavari01** [comment]
same issue

**Comment by ChaseKnowlden** [comment]
Same problem on my website that i'm making.

**Comment by ChaseKnowlden** [comment]
Also happens with up-to-date next.js.

**Comment by nicholasitnikov** [comment]
Same issue

**Comment by amannn** [comment]
I think the Next.js team is aware of the issue, commenting with "same issue" creates a lot of notifications for everyone who is subscribed to this thread and doesn't add anything meaningful. I think leaving a 👍 at the top comment is sufficient!

**Comment by leerob** [comment]
We are planning to address this soon, thank you for the additional reproductions and details here. We hear you all–feel free to upvote the main issue if you are seeing this. We will get this addressed. Thank you 🙏

**Comment by solomonmulatu** [comment]
same issue

**Comment by psikoi** [comment]
I believe this has been fixed with `scroll={false}` in https://github.com/vercel/next.js/pull/51869

Source: https://github.com/vercel/next.js/issues/50105#issuecomment-1619782873

**Comment by ChaseKnowlden** [comment]
Can we close this issue because #51869 is merged, and #50105 is closed?

**Comment by steadyGuy** [comment]
> Can we close this issue because #51869 is merged, and #50105 is closed?

But what about changing query params using _router.push()_? As far as I remember, there was a possibility to provide scroll: true prop, but now it's gone.

Edit:
Oh, I see. They have added this functionality too. Great news!

**Comment by timneutkens** [comment]
While `scroll={false}` allows you to stop scrolling manually there's another case to be solved still, which is the case of searchParams that are not read in server components changing, we still have to add something for that but it's going to take a bit more time as that depends on a new feature we're going to add. We're first focusing on fixing bugs / development performance.

For now you can use `<Link href="/" scroll={false}>` or `router.push('/', { scroll: false })` (`router.replace` works too)

**Comment by exponent28** [comment]
Was experiencing this issue with the new `app` `useRouter` where I couldn't replace the old `{ scroll: false }` behaviour, quite frustrating. For now @YoussefKababe's idea works great.

Had quite a lot of old behaviour to replace so here's a reusable hook: 

**Edit: expose `saveScrollPosition`

```Typescript
// utils/usePersistentScroll.tsx
import { usePathname, useRouter, useSearchParams } from 'next/navigation'
import { useCallback, useEffect } from 'react'

export const usePersistentScroll = () => {
  const router = useRouter()
  const searchParams = useSearchParams()
  const pathname = usePathname()

  useEffect(() => {
    const persistentScroll = localStorage.getItem('persistentScroll')
    if (persistentScroll === null) return

    window.scrollTo({ top: Number(persistentScroll) })

    if (Number(persistentScroll) === window.scrollY) localStorage.removeItem('persistentScroll')
  }, [searchParams])

  const saveScrollPosition = useCallback(() => {
    localStorage.setItem('persistentScroll', window.scrollY.toString())
  }, [])

  const setSearchParam = useCallback(
    (key: string, value: string) => {
      const currentParams = searchParams?.toString()
      const params = new URLSearchParams(currentParams)

      params.set(key, value)

      if (currentParams === params.toString()) return

      saveScrollPosition()
      router.push(`${pathname}?${params.toString()}`)
    },
    [searchParams, pathname, router, saveScrollPosition],
  )

  return { setSearchParam, saveScrollPosition }
}

```

Usage: 
```Typescript
// components/ExampleComponent.tsx
import { usePersistentScroll } from '@/utils/use-persistent-scroll'

export default function ExampleComponent() {
 const { setSearchParam, saveScrollPosition } = usePersistentScroll()

return (
 <button
      onClick={() => {
        setSearchParam('exampleParam', '9999') // or 'saveScrollPosition' for persisting scrollY pos only
      }}
    >
      Update search params
    </button>
)

```

**Comment by aelsaman** [comment]
> 

`router.push(newPathname, {scroll: false});` is still causing a **client-side** page to scroll all the way up.

I test after upgrading to v13.4.9

**Comment by timneutkens** [comment]
@aelsaman reproduction please?

**Comment by aelsaman** [comment]
@timneutkens  well I am not sure if this bug is only related to the app router. However, I am able to reproduce it using the latest version of NextJS `v13.4.9` but using the **pages router**. Here is a simple reproduction steps.

I created a draft new project using npm rather than pnpm, which I was using in my other project, just to exclude any unrelated factors.

<img width="597" alt="CleanShot 2023-07-10 at 14 15 41@2x" src="https://github.com/vercel/next.js/assets/1544585/63a5e10c-af80-4ee6-a70d-eaf98d46658a">

and then, added a simple button in the pages/index page `pages/index.tsx` with dummy router navigation logic as below and it is reproducible for me.

```
<div className='mt-12 h-[1200px]'>page height</div>
<div className="my-4">
  <button
    onClick={() => {
      router.push(`${window.location.pathname}?key1=${new Date().getSeconds()}`, {scroll: false});
    }}
    className="border-2 border-gray-400 p-2 text-lg rounded-md w-32">
    test router
  </button>
</div>
```

It is worth mentioning that the scrolling fix is working as intended if I removed the dynamic value "date seconds" above but it is unstable for me.

Appreciate your help

**Comment by timneutkens** [comment]
@aelsaman so you're using `next/navigation` in `pages`? That'll be fixed by #52498 coincidentally someone just opened that PR.

**Comment by RanaMuzamal** [comment]
Getting the same issue, there is an option that I move the server-side logic to the client-side which was provided by @adrianhajdin in the tutorial, but I don't want to do this. Wanna a solution with server-side rendering? if anyone can guide.

**Comment by skardtechDeveloper** [comment]
Customer key for IMAGIN.studio API is not working

**Comment by RanaMuzamal** [comment]
same here ,is there any other third party API we can use?

**Comment by lmatteis** [comment]
I was able to fix this by adding an empty string (yes an empty string!) as the first child of the container that scrolls. Here's a repo forked by another repo (found in this issue https://github.com/vercel/next.js/issues/43552) that fixes the issue https://github.com/alvarlagerlof/next-app-server-infinite-scroll/pull/160

I have no idea how the empty string as first child causes it to work but my intuition leads me to think there's something with Next's + React diffing that treats the empty string as some sort of placeholder during diffing? Not sure, and indeed quite hacky, but it works for now 😕

**Comment by Olacdy** [comment]
@lmatteis wow, great fix, this actually works as intended when an empty string is added to a DOM.

**Comment by JayvardhanGholap** [comment]
> @timneutkens well I am not sure if this bug is only related to the app router. However, I am able to reproduce it using the latest version of NextJS `v13.4.9` but using the **pages router**. Here is a simple reproduction steps.
> 
> I created a draft new project using npm rather than pnpm, which I was using in my other project, just to exclude any unrelated factors.
> 
> <img alt="CleanShot 2023-07-10 at 14 15 41@2x" width="597" src="https://user-images.githubusercontent.com/1544585/252325824-63a5e10c-af80-4ee6-a70d-eaf98d46658a.png">
> and then, added a simple button in the pages/index page `pages/index.tsx` with dummy router navigation logic as below and it is reproducible for me.
> 
> ```
> <div className='mt-12 h-[1200px]'>page height</div>
> <div className="my-4">
>   <button
>     onClick={() => {
>       router.push(`${window.location.pathname}?key1=${new Date().getSeconds()}`, {scroll: false});
>     }}
>     className="border-2 border-gray-400 p-2 text-lg rounded-md w-32">
>     test router
>   </button>
> </div>
> ```
> 
> It is worth mentioning that the scrolling fix is working as intended if I removed the dynamic value "date seconds" above but it is unstable for me.
> 
> Appreciate your help

Select NO for ESLint and `src`/ directory

**Comment by jackall3n** [comment]
> @lmatteis wow, great fix, this actually works as intended when an empty string is added to a DOM.

 great ~fix~ find

**Comment by zenzen-sol** [comment]
@ryanslee91 This issue relates to app directory structure, which has a different `useRouter` API: https://nextjs.org/docs/app/api-reference/functions/use-router#userouter

**Comment by adrivelasco** [comment]
`router.push('/', { scroll: false });` doesn't work as expected. It's still causing a scroll to top.
Adding an empty string fixes it though.

any progress on this?

**Comment by timneutkens** [comment]
> doesn't work as expected. It's still causing a scroll to top. Adding an empty string fixes it though.

Please provide a reproduction.

**Comment by lmatteis** [comment]
@timneutkens 

> Please provide a reproduction.

https://github.com/vercel/next.js/issues/49087#issuecomment-1639868424

**Comment by timneutkens** [comment]
@lmatteis I'm guessing you're referring to the reproduction for the scroll being reset when searchParams change. That one I can easily reproduce as of right now it's expected behavior as I shared in https://github.com/vercel/next.js/issues/49087#issuecomment-1621201693. The comment linked to does not have `scroll={false}` / `scroll: false` anywhere. @adrivelasco is claiming `router.push` with `scroll: false` doesn't work, that's what I'd like to see a reproduction of.

**Comment by adrivelasco** [comment]
> @lmatteis I'm guessing you're referring to the reproduction for the scroll being reset when searchParams change. That one I can easily reproduce as of right now it's expected behavior as I shared in [#49087 (comment)](https://github.com/vercel/next.js/issues/49087#issuecomment-1621201693). The comment linked to does not have `scroll={false}` / `scroll: false` anywhere. @adrivelasco is claiming `router.push` with `scroll: false` doesn't work, that's what I'd like to see a reproduction of.

oh I thought `router.push('/', { scroll: false });` was what fixes the issue of: scroll being reset when searchParams change. per what you said, scroll being reset is an expected behavior?

+info: I'm using [use-query-params](https://github.com/pbeshai/use-query-params) with a Next.js adapter. I was able to fix this bug only by this [workaround](https://github.com/vercel/next.js/issues/49087#issuecomment-1639868424).

**Comment by rupin27** [comment]
@adrianhajdin
Adding {scroll: false} fixed it for me. Here is my package.json file:

In `router.push(newPathName, {scroll: false});`

package.json:
```
{
  "name": "XYZ",
  "version": "0.1.0",
  "private": true,
  "scripts": {
    "dev": "next dev",
    "build": "next build",
    "start": "next start",
    "lint": "next lint"
  },
  "dependencies": {
    "@headlessui/react": "^1.7.16",
    "@types/node": "20.4.5",
    "@types/react": "18.2.17",
    "@types/react-dom": "18.2.7",
    "autoprefixer": "10.4.14",
    "next": "13.4.12",
    "postcss": "8.4.27",
    "react": "18.2.0",
    "react-dom": "18.2.0",
    "tailwindcss": "3.3.3",
    "typescript": "5.1.6"
  }
}
```

**Comment by ShivangKakkar** [comment]
> @adrianhajdin Adding {scroll: false} fixed it for me. Here is my package.json file:
> 
> In `router.push(newPathName, {scroll: false});`
> 
> package.json:
> 
> ```
> {
>   "name": "XYZ",
>   "version": "0.1.0",
>   "private": true,
>   "scripts": {
>     "dev": "next dev",
>     "build": "next build",
>     "start": "next start",
>     "lint": "next lint"
>   },
>   "dependencies": {
>     "@headlessui/react": "^1.7.16",
>     "@types/node": "20.4.5",
>     "@types/react": "18.2.17",
>     "@types/react-dom": "18.2.7",
>     "autoprefixer": "10.4.14",
>     "next": "13.4.12",
>     "postcss": "8.4.27",
>     "react": "18.2.0",
>     "react-dom": "18.2.0",
>     "tailwindcss": "3.3.3",
>     "typescript": "5.1.6"
>   }
> }
> ```

Thanks, that works!

**Comment by momenqudaih** [comment]
Adding {scroll: false} fixed it for me, with Next 13.4.12 release.❤️

In ``` router.push(newPathName, {scroll: false} );```

**Comment by denisonkolling** [comment]
@momenqudaih @ShivangKakkar @rupin27 

Thanks guys, also worked here...  `"next": "13.4.12"` with  `router.push(newPathName, {scroll: false} );`

**Comment by leerob** [comment]
Got some docs going here, thanks folks 🙏 

https://github.com/vercel/next.js/pull/53804

Edit: wrong button, disregard

**Comment by hikiasi** [comment]
> Adding {scroll: false} fixed it for me, with Next 13.4.12 release.❤️
> 
> In ` router.push(newPathName, {scroll: false} );`

Had the same problem on `next 13.4.13`, solved by adding `{scroll: false}`, thanks a lot!

**Comment by SergiiSechka** [comment]
> I was able to fix this by adding an empty string (yes an empty string!) as the first child of the container that scrolls. Here's a repo forked by another repo (found in this issue #43552) that fixes the issue [alvarlagerlof/next-app-server-infinite-scroll#160](https://github.com/alvarlagerlof/next-app-server-infinite-scroll/pull/160)
> 
> I have no idea how the empty string as first child causes it to work but my intuition leads me to think there's something with Next's + React diffing that treats the empty string as some sort of placeholder during diffing? Not sure, and indeed quite hacky, but it works for now 😕

_________---------------------______________--------------------_________________----------------------__________

where did you add empty string? I tried - it doesn't work for me, could you please make some sort of screenshot or something like that. Thank you

**Comment by SergiiSechka** [comment]
> `router.push('/', { scroll: false });` doesn't work as expected. It's still causing a scroll to top. Adding an empty string fixes it though.
> 
> any progress on this?

{scroll: false} worked for me
Where do you add empty string? can you please post a screenshot how it looks like? Thank you

**Comment by HaKiLo91** [comment]
![image](https://github.com/vercel/next.js/assets/116537483/1ade791c-ad08-4398-8054-04a0306311b7)
performance doesn't seem to be optimal, can you please guide?

**Comment by Zukhriddinbek-Code** [comment]
I also have same issue, unfortunately in a project we have to switch from server to client side

**Comment by Achour** [comment]
> I'm experiencing the same bug. The website scrolls to the top every time the URL parameters change.
> 
> I saw @timneutkens's [tweet](https://twitter.com/timneutkens/status/1645138528781709314?s=20) on the topic.
> 
>  2023-05-27.09-58-58.mp4
try this: 
router.push(newPathName, { scroll: false })
it will fix the issue

**Comment by alexhollender** [comment]
**version:** `"next": "13.4.19"`
**router**: `app router`
**solution:** adding `{ scroll: false }` to my `router.push()` call fixed the issue, for example:
```
router.push(`${pathname}${updatedSearchParams}`, {scroll: false});
```
**documentation:** https://nextjs.org/docs/app/building-your-application/routing/linking-and-navigating#disabling-scroll-restoration

thanks @timneutkens for the solution

**Comment by Mishael-Joe** [comment]
@ShivangKakkar thanks too
@adrianhajdin Adding {scroll: false} fixed it for me. Here is a screenshot of the solution:

In router.push(newPathName, {scroll: false});

![searchparams](https://github.com/vercel/next.js/assets/116783000/3659877f-81c7-4293-8195-0c1949394354)

**Comment by crisner1978** [comment]
you can set the scroll options and this will prevent the scrolling when trying to change your searchParams

router.push(newPathname, { scroll: false})

**Comment by timneutkens** [comment]
I'm going to lock this issue as it's already part of our issue tracker and the same solution keeps being repeated which causes notifications to everyone in this thread.

**Comment by samcx** [comment]
Hi everyone!

I will be moving this to <picture data-single-emoji=":nextjs:" title=":nextjs:"><img class="emoji" width="20" height="auto" src="https://emoji.slack-edge.com/T0CAQ00TU/nextjs/2a85d633ae594556.png" alt=":nextjs:" align="absmiddle"></picture> [Discussions](https://github.com/vercel/next.js/discussions) as this is no longer an active issue (locked), and the solution is shared above.

Happy 2024!
